### PR TITLE
[SPARK-17494] [SQL] changePrecision() on compact decimal should respect rounding mode

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

Floor()/Ceil() of decimal is implemented using changePrecision() by passing a rounding mode, but the rounding mode is not respected when the decimal is in compact mode (could fit within a Long). 

This Update the changePrecision() to respect rounding mode, which could be ROUND_FLOOR, ROUND_CEIL, ROUND_HALF_UP, ROUND_HALF_EVEN.
## How was this patch tested?

Added regression tests.

### CODE REVIEW

**Comment by davies** [comment]
cc @JoshRosen @chenghao-intel

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 1]
```diff
@@ -242,10 +242,30 @@ final class Decimal extends Ordered[Decimal] with Serializable {
...
+            if (droppedDigits < 0) {
+              longVal += -1L
+            }
+          case ROUND_CEILING =>
+            if (droppedDigits > 0) {
+              longVal += 1L
+            }
+          case ROUND_HALF_UP =>
+            if (math.abs(droppedDigits) * 2 >= pow10diff) {
+              longVal += (if (longVal < 0) -1L else 1L)
```
Comment (srinathshankar) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: Is this correct if longVal is 0, after the division above
What happens if you round_half_up -0.6 ? Looks like you may get +1 ?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/types/DecimalSuite.scala [hunk 2]
```diff
@@ -191,4 +192,16 @@ class DecimalSuite extends SparkFunSuite with PrivateMethodTester {
     assert(new Decimal().set(100L, 10, 0).toUnscaledLong === 100L)
     assert(Decimal(Long.MaxValue, 100, 0).toUnscaledLong === Long.MaxValue)
   }
+
+  test("changePrecision() on compact decimal should respect rounding mode") {
+    Seq(ROUND_FLOOR, ROUND_CEILING, ROUND_HALF_UP, ROUND_HALF_EVEN).foreach { mode =>
+      Seq("1.0", "1.1", "1.6", "2.5", "5.5", "-1.0", "-1.1", "-1.6", "-2.5", "-5.5").foreach { n =>
```
Comment (srinathshankar) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/types/DecimalSuite.scala]: tests of the form 0.5, -0.5, 0.6 etc. might also be useful

**Review by srinathshankar — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/types/DecimalSuite.scala (hunk 2 above)
Comment (davies) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/types/DecimalSuite.scala]: Currently, these numbers are copied from the java doc of RoundingMode, will add more for ROUND_HALF_EVEN

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala (hunk 1 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: Good catch, we should use droppedDigits here.

**Review by davies — COMMENTED** [review]
Thanks for the comment, pushed the fix

File: sql/catalyst/src/test/scala/org/apache/spark/sql/types/DecimalSuite.scala [hunk 3]
```diff
@@ -191,4 +192,18 @@ class DecimalSuite extends SparkFunSuite with PrivateMethodTester {
     assert(new Decimal().set(100L, 10, 0).toUnscaledLong === 100L)
     assert(Decimal(Long.MaxValue, 100, 0).toUnscaledLong === Long.MaxValue)
   }
+
+  test("changePrecision() on compact decimal should respect rounding mode") {
+    Seq(ROUND_FLOOR, ROUND_CEILING, ROUND_HALF_UP, ROUND_HALF_EVEN).foreach { mode =>
+      Seq("0.4", "0.5", "0.6", "1.0", "1.1", "1.6", "2.5", "5.5").foreach { n =>
+        Seq("", "-").foreach { sigh =>
```
Comment (srinathshankar) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/types/DecimalSuite.scala]: I think you mean sign, not sigh, though I can understand :)

**Review by srinathshankar — APPROVED** [review]
LGTM.

**Comment by rxin** [comment]
Merging in master/2.0.

**Pull request closed without merging**